File: scripts/generate-favicon.py
from __future__ import annotations

import json
from pathlib import Path

from PIL import Image


PROJECT_ROOT = Path(__file__).resolve().parents[1]
INPUT_PNG = PROJECT_ROOT / "public" / "appicon.png"

OUTPUTS = {
    "png16": PROJECT_ROOT / "favicon-16x16.png",
    "png32": PROJECT_ROOT / "favicon-32x32.png",
    "apple180": PROJECT_ROOT / "apple-touch-icon.png",
    "ico": PROJECT_ROOT / "favicon.ico",
    "manifest": PROJECT_ROOT / "site.webmanifest",
}


def resize_square(img: Image.Image, size: int) -> Image.Image:
    # Ensure RGBA so transparent backgrounds stay transparent
    img = img.convert("RGBA")

    # If it's not square, contain it within a square canvas
    w, h = img.size
    if w != h:
        side = max(w, h)
        canvas = Image.new("RGBA", (side, side), (0, 0, 0, 0))
        canvas.paste(img, ((side - w) // 2, (side - h) // 2))
        img = canvas

    return img.resize((size, size), resample=Image.Resampling.LANCZOS)


def main() -> None:
    if not INPUT_PNG.exists():
        raise SystemExit(f"Missing input file: {INPUT_PNG}")

    base = Image.open(INPUT_PNG)

    resize_square(base, 16).save(OUTPUTS["png16"], format="PNG", optimize=True)
    resize_square(base, 32).save(OUTPUTS["png32"], format="PNG", optimize=True)
    resize_square(base, 180).save(OUTPUTS["apple180"], format="PNG", optimize=True)

    # Generate an .ico with multiple embedded sizes for compatibility
    ico_sizes = [(16, 16), (24, 24), (32, 32), (48, 48), (64, 64), (128, 128), (256, 256)]
    base_rgba = resize_square(base, 256)
    base_rgba.save(OUTPUTS["ico"], format="ICO", sizes=ico_sizes)

    manifest = {
        "name": "TAU Icon Generator",
        "short_name": "TAU",
        "icons": [
            {"src": "favicon-16x16.png", "sizes": "16x16", "type": "image/png"},
            {"src": "favicon-32x32.png", "sizes": "32x32", "type": "image/png"},
            {"src": "apple-touch-icon.png", "sizes": "180x180", "type": "image/png"},
        ],
        "display": "standalone",
    }
    OUTPUTS["manifest"].write_text(json.dumps(manifest, indent=2) + "\n", encoding="utf-8")

    print("Generated:")
    for key in ["ico", "png16", "png32", "apple180", "manifest"]:
        print(f"- {OUTPUTS[key].relative_to(PROJECT_ROOT)}")


if __name__ == "__main__":
    main()

File: scripts/generate-favicon.mjs
import fs from "node:fs/promises"
import path from "node:path"
import sharp from "sharp"
import pngToIco from "png-to-ico"

const projectRoot = path.resolve(process.cwd())
const inputPng = path.join(projectRoot, "public", "appicon.png")

const outputs = {
  png16: path.join(projectRoot, "favicon-16x16.png"),
  png32: path.join(projectRoot, "favicon-32x32.png"),
  apple180: path.join(projectRoot, "apple-touch-icon.png"),
  ico: path.join(projectRoot, "favicon.ico"),
  manifest: path.join(projectRoot, "site.webmanifest"),
}

async function fileExists(filePath) {
  try {
    await fs.stat(filePath)
    return true
  } catch {
    return false
  }
}

async function main() {
  if (!(await fileExists(inputPng))) {
    console.error(`Missing input: ${inputPng}`)
    process.exit(1)
  }

  const base = sharp(inputPng).ensureAlpha().resize(512, 512, { fit: "contain" })

  await base.clone().resize(16, 16).png().toFile(outputs.png16)
  await base.clone().resize(32, 32).png().toFile(outputs.png32)
  await base
    .clone()
    .resize(180, 180)
    .png({ compressionLevel: 9 })
    .toFile(outputs.apple180)

  // favicon.ico: embed multiple sizes for best compatibility
  const icoPngBuffers = await Promise.all(
    [16, 24, 32, 48, 64, 128, 256].map((size) => base.clone().resize(size, size).png().toBuffer()),
  )

  const icoBuffer = await pngToIco(icoPngBuffers)
  await fs.writeFile(outputs.ico, icoBuffer)

  const manifest = {
    name: "TAU Icon Generator",
    short_name: "TAU",
    icons: [
      { src: "favicon-16x16.png", sizes: "16x16", type: "image/png" },
      { src: "favicon-32x32.png", sizes: "32x32", type: "image/png" },
      { src: "apple-touch-icon.png", sizes: "180x180", type: "image/png" },
    ],
    display: "standalone",
  }
  await fs.writeFile(outputs.manifest, JSON.stringify(manifest, null, 2) + "\n", "utf8")

  console.log("Generated:")
  console.log("- favicon.ico")
  console.log("- favicon-16x16.png")
  console.log("- favicon-32x32.png")
  console.log("- apple-touch-icon.png")
  console.log("- site.webmanifest")
}

main().catch((err) => {
  console.error(err)
  process.exit(1)
})
